src/efftester.ml

```diff
@@ -290,275 +304,328 @@ exception No_solution
-    List etyp in
+        List.fold_left
+          (fun typacc elem ->
+            let etyp = lit_type elem in
+            if types_compat typacc etyp (* if typacc is a generalization of etyp *)
+            then etyp
+            else if types_compat etyp typacc (* if etyp is a generalization of typeacc *)
+            then typacc
+            else
+              failwith
+                ("lit_type: elements in list literal disagree"
+                ^ "  typacc is "
+                ^ typetostr ~effannot:true typacc
+                ^ "  etyp is "
+                ^ typetostr ~effannot:true etyp))
+          (Typevar (newtypevar ()))
+          l
+      in
+      List etyp
+  in
   match t with
   | Lit l -> lit_type l
-  | Variable (typ,_)    -> typ
-  | Lambda (typ,_,_,_)  -> typ
-  | App (typ,_,_,_,_)   -> typ
-  | Let (_,_,_,_,typ,_) -> typ
-  | If (typ,_,_,_,_)    -> typ
-    
+  | Variable (typ, _) -> typ
+  | Lambda (typ, _, _, _) -> typ
+  | App (typ, _, _, _, _) -> typ
+  | Let (_, _, _, _, typ, _) -> typ
+  | If (typ, _, _, _, _) -> typ
+;;
 
 (** Enviroment type definitions and functions *)
 
-module TypeSet = Set.Make(struct
-                           type t = etype
-			   let compare = Pervasives.compare
-                         end)
-module VarSet = Set.Make(struct
-                           type t = variable
-			   let compare = String.compare
-                         end)
-module VarMap = Map.Make(struct
-                           type t = variable
-			   let compare = String.compare
-                         end)
-module TypeMap = Map.Make(struct
-                           type t = etype
-			   let compare = Pervasives.compare
-                          end)
-
-let rec normalize_eff t = match t with
-  | Typevar _
-  | Unit
-  | Int
-  | Bool
-  | String -> t
-  | List t' -> let t'' = normalize_eff t' in
-	       List t''
-  | Fun (s,e,t) -> Fun (normalize_eff s,
-			no_eff,
-			normalize_eff t)
-    
-let addMultiMap key value map = (* assume key has been normalized *)
+module TypeSet = Set.Make (struct
+  type t = etype
+
+  let compare = Pervasives.compare
+end)
+
+module VarSet = Set.Make (struct
+  type t = variable
+
+  let compare = String.compare
+end)
+
+module VarMap = Map.Make (struct
+  type t = variable
+
+  let compare = String.compare
+end)
+
+module TypeMap = Map.Make (struct
+  type t = etype
+
+  let compare = Pervasives.compare
+end)
+
+let rec normalize_eff t =
+  match t with
+  | Typevar _ | Unit | Int | Bool | String -> t
+  | List t' ->
+    let t'' = normalize_eff t' in
+    List t''
+  | Fun (s, e, t) -> Fun (normalize_eff s, no_eff, normalize_eff t)
+;;
+
+let addMultiMap key value map =
+  (* assume key has been normalized *)
   try
     let s = TypeMap.find key map in
     let newSet = VarSet.add value s in
     TypeMap.add key newSet map
-  with Not_found ->
-    TypeMap.add key (VarSet.singleton value) map
+  with
+  | Not_found -> TypeMap.add key (VarSet.singleton value) map
```

Comment:
I am not sure what you mean here

Reply:
This is not a formatting comment but more of a refactoring comment: whenever you see a pattern of the form

```ocaml
try
  (* some amount of code *)
with exn -> (* exception handling code *)
```

it is useful to wonder: in the `try .. with` block, what part of the code may actually raise the exception being handled? If there are several different parts, or if it's the whole expression that may raise, then `try .. with` is the right construct to use. If there is only one sub-part of the code that may raise the exception, then it is usually nicer to rewrite the code as

```ocaml
match <part that may raise> with
| exception exn -> (* the exception-handling code *)
| result -> (* the rest of the code *)
```

In the present case, this would suggest the following:

```ocaml
match TypeMap.find key map with
| exception Not_found -> ...
| s -> ...
```

(On the ordering of the two clauses (the value case, the exception case), I usually put whichever is shortest first, with a preference to the exceptional case in case of tie.)

Reply:
That indeed makes code more readable. I haven't been familiar with this OCaml feature before. Thanks! :-)